From a4ee487213d3820c5c664c153103425c31cf3ca2 Mon Sep 17 00:00:00 2001
From: Dmitry Torokhov <dtor@chromium.org>
Date: Mon, 27 Apr 2015 15:48:02 -0700
Subject: [PATCH] CHROMIUM: Input: elants_i2c - keep regulators on when
 unbinding

Until we come up with a good story for controlling regulators from
userspace let's keep them on when unbinding drivers so that factory
test utilities can continue accessing devices via i2c-dev on ARM in the
same fashion as they do that on X86.

BUG=chrome-os-partner:39363
TEST=Build and boot Minnie with Elan touchscreen, unbind the touchscreen
via sysfs and use i2cdetect to verify that device is still powered up
and responding to queries.

Change-Id: I061af63d0c91dc5c5a90df6d43ff021e4ad51ce7
Signed-off-by: Dmitry Torokhov <dtor@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/267582
Reviewed-by: Douglas Anderson <dianders@chromium.org>
Trybot-Ready: Douglas Anderson <dianders@chromium.org>
Tested-by: Jiazi Yang <Tomato_Yang@asus.com>
Commit-Queue: Douglas Anderson <dianders@chromium.org>
(cherry picked from commit 7213eb4731dbbb77317494a0b0fcdda162969157)
Reviewed-on: https://chromium-review.googlesource.com/354085

[rebase419(rrangel): Context conflict]
Signed-off-by: Raul E Rangel <rrangel@chromium.org>
---
 drivers/input/touchscreen/elants_i2c.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/input/touchscreen/elants_i2c.c b/drivers/input/touchscreen/elants_i2c.c
index f2cb231218339..325ae6054d62a 100644
--- a/drivers/input/touchscreen/elants_i2c.c
+++ b/drivers/input/touchscreen/elants_i2c.c
@@ -149,6 +149,7 @@ struct elants_data {
 
 	bool wake_irq_enabled;
 	bool keep_power_in_suspend;
+	bool unbinding;
 
 	/* Must be last to be used for DMA operations */
 	u8 buf[MAX_PACKET_SIZE] ____cacheline_aligned;
@@ -1121,6 +1122,12 @@ static void elants_i2c_power_off(void *_data)
 {
 	struct elants_data *ts = _data;
 
+	if (ts->unbinding) {
+		dev_info(&ts->client->dev,
+			 "Not disabling regulators to continue allowing userspace i2c-dev access\n");
+		return;
+	}
+
 	if (!IS_ERR_OR_NULL(ts->reset_gpio)) {
 		/*
 		 * Activate reset gpio to prevent leakage through the
@@ -1296,6 +1303,19 @@ static int elants_i2c_probe(struct i2c_client *client,
 	return 0;
 }
 
+static int elants_i2c_remove(struct i2c_client *client)
+{
+	struct elants_data *ts = i2c_get_clientdata(client);
+
+	/*
+	 * Let elants_i2c_power_off know that it needs to keep
+	 * regulators on.
+	 */
+	ts->unbinding = true;
+
+	return 0;
+}
+
 static int __maybe_unused elants_i2c_suspend(struct device *dev)
 {
 	struct i2c_client *client = to_i2c_client(dev);
@@ -1393,6 +1413,7 @@ MODULE_DEVICE_TABLE(of, elants_of_match);
 
 static struct i2c_driver elants_i2c_driver = {
 	.probe = elants_i2c_probe,
+	.remove = elants_i2c_remove,
 	.id_table = elants_i2c_id,
 	.driver = {
 		.name = DEVICE_NAME,
-- 
2.22.0.770.g0f2c4a37fd-goog

